### ISSUE TITLE
How to have an output of the CoralOrdinal layer to be a single integer of predicted class?

### ISSUE THREAD

**Issue by marftabucks** [ck37/coral-ordinal#22]
Hello.

I found out that the ordinal softmax function yields an output of list which contains NUM_CLASSES - 1. In other words, it's a list. I have been wondering how to have the layer to give an output of the predicted class in an integer. I found some activation functions in the activations module: 1) coral_cumprobs and 2) cumprobs_to_label. 

Correct me if I'm wrong. If my goal is to have a single integer of predicted class in the output. I could use both functions mentioned above in order. Hence, I tried to created a custom activation function, something like this

```
def coral_labels(x: tf.Tensor):
  threshold = 0.5

  result = tf.math.sigmoid([x])
  predict_levels = tf.cast(result > threshold, dtype=tf.int32)
  predicted_labels = tf.reduce_sum(predict_levels, axis=1)
  return predicted_labels
```

However, the model summary still tells that the output layer is a list of (1, NUM_CLASSES - 1)

<img width="504" alt="image" src="https://user-images.githubusercontent.com/55443235/177037055-4d31058a-7c4c-4b6d-8447-b9219c4076b7.png">

Could you help me on this?

**Comment by gmgeorg** [comment]
@marftabucks I think it's useful to review prediction vs classification.  For example, have a look at
https://stats.stackexchange.com/questions/464636/proper-scoring-rule-when-there-is-a-decision-to-make-e-g-spam-vs-ham-email/538442?noredirect=1#comment1107477_538442
or
https://www.fharrell.com/post/classification/

Particular to your question here, ordinal regression (not classification!) gives you estimates of the Pr(label > i).  It is up to the decision maker -- you or your stakeholder -- to take those probabilities and turn them into a classification of the decision (label) you want to pick.  There is no single *best* way to do that; it depends on the cost of making errors. 

Now you *could* write a custom layer that turns ordinal probabilities into integers (e.g., ordinal probs -> softmax probs -> tf.argmax()), add it to the model you have, and then adapt the loss function / metric that works off labels (e.g., precision).  However, as explained in the posts above, this is usually *not* recommended, as you combine prediction & classification in a single process, which makes it much harder to use in practice.  It also makes it impossible to differentiate model predictive performance from the decision making process.

See in particular the https://github.com/ck37/coral-ordinal/blob/master/coral_ordinal/activations.py , with all the helper functions that turn CORN/CORAL probabilities into softmax and labels.

**Comment by haoyang27** [comment]
Hi,

I'm a beginner on deep learning and have been trying to turn a multi-class classification task into the ordinal regression. Here's the code regarding the CNN architecture:
`from tensorflow import keras
import coral_ordinal as coral
import tensorflow as tf

# This is a list of all possible models to create
model_list = {
    "resnet_50": keras.applications.resnet50.ResNet50, 
    "resnet_VGG16": keras.applications.VGG16, 
    "InceptionV3": keras.applications.InceptionV3,
    "ResNet50V2": keras.applications.ResNet50V2, 
    "Xception": keras.applications.Xception
}


class TrainingModel:
    def __init__(self, hyperparameters, model_type, target_height, target_width, class_names):
        """ Creates and prepares a model for training.
            
        Args:
            hyperparameters (dict): The configuration's hyperparameters.
            model_type (str): Type of model to create.
            target_height (int): Height of input.
            target_width (int): Width of input.
            class_names (list of str): A list of classes.
            
        Returns:
            model (keras Model): The prepared keras model.
            model_type (str): The name of the model type.
        """

        # Catch if the model is not in the model list
        if model_type not in model_list:
            print(colored(f"Warning: Model '{model_type}' not found in the list of possible models: {list(model_list.keys())}"))
            
            # TODO: Create a custom model

        self.model_type = model_type
            
        # Get the model base
        base_model = model_list[model_type](
            include_top=False,
            weights=None,
            input_tensor=None,
            input_shape=(target_height, target_width, hyperparameters['channels']),
            pooling=None
        )

        # Return the prepared model
        avg = keras.layers.GlobalAveragePooling2D()(base_model.output)
        #out = coral.CoralOrdinal(len(class_names))(avg)
        out = keras.layers.Dense(len(class_names), activation="softmax")(avg)
        out = coral.CoralOrdinal(len(class_names))(out)
        self.model = keras.models.Model(inputs=base_model.input, outputs=out)
        
        # Create optimizer and add to model
        optimizer = keras.optimizers.SGD(
            lr=hyperparameters['learning_rate'], 
            momentum=hyperparameters['momentum'], 
            nesterov=True, 
            decay=hyperparameters['decay']
        )
        self.model.compile(
            #loss="sparse_categorical_crossentropy", 
            loss = coral.OrdinalCrossEntropy(),
            optimizer=optimizer,
            #metrics=["accuracy"]
            metrics = ["accuracy",coral.MeanAbsoluteErrorLabels()]       
        )`

Am I doing this correctly by adding a ordinal layer, changing the loss function and the metrics? Also, does the accuracy actually make sense since I'm getting extremely low value on accuracy(around 0.05, it's around 0.95 when it's a classification). Any help would be greatly appreciated, thanks!!!
